### PR TITLE
ISSUE-292 Prevent SPI calls at runtime

### PR DESCRIPTION
**Pull request by chibenwa**
Closes https://github.com/opensearch-project/opensearch-java/issues/292

### CODE REVIEW

**Comment by wbeckler** [comment]
What's the issue this addresses?

**Comment by wbeckler** [comment]
Ah I see it here: https://github.com/opensearch-project/opensearch-java/issues/292

File: java-client/src/main/java/org/opensearch/client/json/jackson/JsonValueParser.java [hunk 1]
```diff
@@ -49,7 +49,7 @@
  * object (e.g. START_OBJECT, VALUE_NUMBER, etc).
  */
 class JsonValueParser {
-    private final JsonProvider provider = JsonProvider.provider();
+    private static final JsonProvider provider = JsonProvider.provider();
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/json/jackson/JsonValueParser.java:52]: Let's postpone the provider initialization till it is really needed? 

```suggestion
    private static class DefaultJsonProvider {
        private static final JsonProvider INSTANCE = JsonProvider.provider();
    }    
```

And than just replace `provider` with `DefaultJsonProvider.INSTANCE` ?

**Comment by dblock** [comment]
Can we please add a unit test for this, too?

**Comment by wbeckler** [comment]
I think it makes sense that this would go testless. On the other hand, it
would be awesome if you could open an issue suggesting an ongoing approach
to performance, to perhaps detect other low hanging fruit.

**Comment by chibenwa** [comment]
Next week i'll propose a JMH bench for this changeset.

Regarding "detection of low hanging fruits" I use asyncprofiler on my workloads (memory & cpu). Can't recall other massive painpoints but I will get a second look, and share flame graphs here too.

Happy new year,

Regards,

Benoit

**Comment by chibenwa** [comment]
[opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip)

Here is the flame graph I diagnosed this issue from. Taken with https://github.com/jvm-profiling-tools/async-profiler :

```
./profiler -d 120 -e itimer -f opensearch_2.4.0.html 1
```

(From within my applicative container)

Then searching `opensearch` (top left) gives you:
  - The driver thread (on the right) with the httpclient event loop
  - Applicatives calls to the driver from applicative threads (small stacktraces in the center)
  
Observations for the driver event loop:
  - Call to the SPI are very easy to notice CF issue. Not only it eats CPU/heap allocations but also does it blocks the HTTP event loop, which is catastrophic (my app is not OpenSearch heavy so that is not a big issue to me).
  - 0.5% of my application CPU (and 5% of the eventLoop CPU - similar footprint on heap allocation) is eaten by HTTP-client basic auth caching... :-( **Likely a low hanging fruit, shall I create an issue for this one?**
  - Unsurprisingly JSON parsing takes ~11% of the drivers CPU and 24% of heap allocation.
  - Event loop busyness (push stuff to the kernel network stack, SSL) takes around 60% of the event loop CPU, 25% of heap allocations, which again feels normal to me.
  - 2.3% of the event loop CPU is our binding to Reactor reactive library, which, again is normal: converting futures and enqueuing tasks takes time.

**Comment by chibenwa** [comment]
## JMH output backing this change

### Before

```
Benchmark                Mode  Cnt   Score   Error  Units
JMHFieldBench.jsonBench  avgt    5  71.790 ± 2.125  us/op
```

### After

```
Benchmark                Mode  Cnt  Score   Error  Units
JMHFieldBench.jsonBench  avgt    5  1.418 ± 0.024  us/op
```

### Conclusion

x50 fastening of the JSOn parsing by getting rid of the SPI lookups, not event taking into account blocking operations... Huge win!

### Code of the benchmark

```
package org.apache.james.backends.opensearch;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import org.apache.james.backends.opensearch.json.jackson.JacksonJsonpParser;
import org.junit.Test;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.infra.Blackhole;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;
import org.openjdk.jmh.runner.options.TimeValue;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.ObjectMapper;

public class JMHFieldBench {

    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    @Test
    public void launchBenchmark() throws Exception {
        Options opt = new OptionsBuilder()
            .include(this.getClass().getName() + ".*")
            .mode (Mode.AverageTime)
            .timeUnit(TimeUnit.MICROSECONDS)
            .warmupTime(TimeValue.seconds(5))
            .warmupIterations(3)
            .measurementTime(TimeValue.seconds(2))
            .measurementIterations(5)
            .threads(1)
            .forks(1)
            .shouldFailOnError(true)
            .shouldDoGC(true)
            .build();

        new Runner(opt).run();
    }

    @Benchmark
    public void jsonBench(Blackhole bh) throws IOException {
        final JsonParser parser = OBJECT_MAPPER.createParser("[\"a\",\"b\"]");
        try {
            bh.consume(new JacksonJsonpParser(parser).getArrayStream());
        } catch (Exception e) {
            // do nothing
        } finally {
            parser.close();
        }
    }
}
```

**Comment by reta** [comment]
It is known that `ObjectMapper` is expensive to create, making it  static certainly helps

**Comment by chibenwa** [comment]
> It is known that ObjectMapper is expensive to create, making it static certainly helps

I do not see how this affects this benchmark

**Comment by reta** [comment]
> > It is known that ObjectMapper is expensive to create, making it static certainly helps
> 
> I do not see how this affects this benchmark

Apologies, your benchmark checks the impact of changing the `JsonProvider.provider()` initialization (instance -> static), for some reason I thought that `JsonProvider.provider()` was instantiating ObjectMapper instances (which is not correct).

**Comment by wbeckler** [comment]
Would you be up for adding a change to the changelog?

**Review by reta — APPROVED** [review]

**Review by dblock — APPROVED** [review]

**Pull request merged**